tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncLineDebuggerInvoker.cs

```diff
@@ -0,0 +1,225 @@
+// <copyright file="AsyncLineDebuggerInvoker.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Debugger.Helpers;
+using Datadog.Trace.Debugger.RateLimiting;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Instrumentation
+{
+    /// <summary>
+    /// LineDebuggerInvoker Invoker
+    /// </summary>
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class AsyncLineDebuggerInvoker
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(LineDebuggerInvoker));
+
+        private static AsyncLineDebuggerState CreateInvalidatedAsyncLineDebuggerState()
+        {
+            var defaultState = AsyncLineDebuggerState.GetDefault();
+            defaultState.IsActive = false;
+            return defaultState;
+        }
+
+        /// <summary>
+        /// Logs the given <paramref name="local"/> ByRef.
+        /// </summary>
+        /// <typeparam name="TLocal">Type of local.</typeparam>
+        /// <param name="local">The local to be logged.</param>
+        /// <param name="index">index of given argument.</param>
+        /// <param name="state">Debugger state</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void LogLocal<TLocal>(ref TLocal local, int index, ref AsyncLineDebuggerState state)
+        {
+            if (!state.IsActive)
+            {
+                return;
+            }
+
+            var localVariableNames = state.MethodMetadataInfo.LocalVariableNames;
+            if (!MethodDebuggerInvoker.TryGetLocalName(index, localVariableNames, out var localName))
+            {
+                return;
+            }
+
+            state.SnapshotCreator.CaptureLocal(local, localName);
+            state.HasLocalsOrReturnValue = true;
+        }
+
+        /// <summary>
+        /// Log exception
+        /// </summary>
+        /// <param name="exception">Exception instance</param>
+        /// <param name="state">Debugger state</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void LogException(Exception exception, ref AsyncLineDebuggerState state)
+        {
+            Log.Warning(exception, "Error caused by our instrumentation");
+            state.IsActive = false;
```

Comment:
We should only write this log warning if `state.IsActive` is `true. Otherwise, if we instrument a method that gets executed a million times, we might end up writing a log a million times which might cause the customer's code to time out.

Reply:
Nice catch. Done in 012c103f14219b94b4f1edafea14e79cfc7f1e54.